### ISSUE TITLE
ipsec: IPv6 reply bypasses tunnel which might lead to unintended masquerading

### ISSUE THREAD

**Issue by brb** [cilium/cilium#23461]
Hit the issue that cilium-health was reporting failing IPv6 host netns -> remote pod checks. Closer look revealed that a second reply from the remote health check pod got masqueraded by Cilium's rule:
```
[302:21728] -A CILIUM_POST_nat ! -s fd02::/120 ! -d fd02::/120 -o cilium_host -m comment --comment "cilium host->cluster masquerade" -j SNAT --to-source fc00:f853:ccd:e793::5
```
According to @pchaigno the reply should not hit the networking stack, and instead it should have redirected to the `cilium_vxlan` bypassing the rule.

To reproduce (taken from https://github.com/cilium/cilium/actions/runs/4024902232/jobs/6917456832):
```
cd cilium/
./contrib/scripts/kind.sh "" 2 "" "" "iptables" "dual"
 helm template --validate ./install/kubernetes/cilium --namespace=kube-system  --set preflight.image.useDigest=false  --set ipv4NativeRoutingCIDR=10.0.0.0/8  --set ipv6NativeRoutingCIDR=fd02::/112  --set operator.image.repository=quay.io/cilium/operator  --set encryption.enabled=true  --set ipv6.enabled=true  --set hubble.relay.image.useDigest=false  --set image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set etcd.leaseTTL=30s  --set logSystemLoad=true  --set debug.enabled=true  --set operator.image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set bandwidthManager.enabled=false  --set hubble.eventBufferCapacity=65535  --set ipam.operator.clusterPoolIPv6PodCIDR=fd02::/112  --set ipv4.enabled=true  --set hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci  --set preflight.image.repository=quay.io/cilium/cilium-ci  --set hubble.listenAddress=:4244  --set bpf.masquerade=false  --set enableCnpStatusUpdates=true  --set hubble.relay.image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set image.useDigest=false  --set image.repository=quay.io/cilium/cilium-ci  --set pprof.enabled=true  --set operator.image.suffix=-ci  --set bpf.preallocateMaps=false  --set sessionAffinity=false  --set hubble.enabled=true  --set operator.image.useDigest=false  --set debug.verbose=flow  --set kubeProxyReplacement=disabled  --set tunnel=vxlan  --set preflight.image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set k8s.requireIPv4PodCIDR=true  --set enableCiliumEndpointSlice=true  > cilium-173e2a16d47c0211.yaml
k apply -f test/k8s/manifests/ipsec_secret.yaml -n kube-system
k apply -f cilium-173e2a16d47c0211.yaml
```
Then exec into any cilium-agent pod, and run:
```
cilium-health status -o json --probe
```

**Comment by pchaigno** [comment]
Interestingly this doesn't affect IPv4 (I would expect it to). The tunnel bypass is from https://github.com/cilium/cilium/blob/5c811513eae2bfe2bf745277f5c7b71b9290ca14/bpf/lib/encap.h#L213-L227.

I have a sysdump in DMs with Martynas from today.

**Comment by brb** [comment]
If it helps, the pwru output for the v4 reply path:
```
0xffff8ad8402ac6e8      0 [cilium-health-r]             nf_hook_slow sec-arg=18446633258184621624 netns=4026534895 mark=0x0 ifindex=19 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r] nft_do_chain_ipv4   [nf_tables] sec-arg=18446615260126955240 netns=4026534895 mark=0x0 ifindex=19 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]        ip_forward_finish sec-arg=0 netns=4026534895 mark=0x0 ifindex=19 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]                ip_output sec-arg=0 netns=4026534895 mark=0x0 ifindex=19 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]             nf_hook_slow sec-arg=18446633258184621616 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r] nft_do_chain_ipv4   [nf_tables] sec-arg=18446615260126955240 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r] nf_nat_ipv4_out     [nf_nat] sec-arg=18446615260126955240 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]  nf_nat_inet_fn     [nf_nat] sec-arg=18446615260126955240 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r] ipv4_confirm        [nf_conntrack] sec-arg=18446615260126955240 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r] nf_confirm  [nf_conntrack] sec-arg=20 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]         ip_finish_output sec-arg=0 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]       __ip_finish_output sec-arg=0 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]        ip_finish_output2 sec-arg=0 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=127 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]         __dev_queue_xmit sec-arg=0 netns=4026534895 mark=0x0 ifindex=3 proto=8 mtu=1500 len=141 10.0.0.126:4240->10.0.2.115:46812(tcp)
0xffff8ad8402ac6e8      0 [cilium-health-r]             tcf_classify sec-arg=0 netns=4026534895 mark=0x0 ifindex=3 proto=8
```

* ifindex=19: `lxc_health`
* ifindex=3: `cilium_host` (which later redirects the packet over the tunnel).

**Comment by jschwinger233** [comment]
Got some progress here.

TL;DR

ip6table misses 2 rules to exclude xfrm marks from nat CILIUM_POST_nat chain.

Long story:

1. iptables command lines seems to be buggy and can't show our excluding rules:
```shell
$ docker exec -it kind-worker2 ip6tables -t nat -L CILIUM_POST_nat --line-numbers
Chain CILIUM_POST_nat (1 references)
num  target     prot opt source               destination         
1    ACCEPT     all      anywhere             anywhere             /* exclude xfrm marks from nat CILIUM_POST_nat chain */
2    ACCEPT     all      anywhere             anywhere             /* exclude xfrm marks from nat CILIUM_POST_nat chain */
```
The output is actually not correct as we set rules like `ip6tables -t nat -A CILIUM_POST_nat -m mark --mark 0x00000d00/0x00000f00 -m comment --comment xxx -j ACCEPT`.

2. For ipsec, `from-container` bpf program will pass the skb to stack with specific marks to hit xfrm policy, but the weird thing is the second tcp ack from health server can't be encapsulated into ESP during the datapath `lxc_health -> stack -> cilium_host`. It turns out it's ip6tables missing these excluding rules causes trouble. Once I add these rules as what we do for ip4table, the tcp session from host1 to health@host2 recovers.

I'm to fix this in the PR https://github.com/isovalent/roadmap/issues/538.

**Comment by pchaigno** [comment]
> I'm to fix this in the PR #24208

Is it possible to move this bug fix to its own pull request or does it depend on other changes in that pull request?

**Comment by jschwinger233** [comment]
> Is it possible to move this bug fix to its own pull request or does it depend on other changes in that pull request?

Just tried to cherry-pick 364fce2942f9f8a2cd509d5e222c56c433880204 and c7c50cb1d79b85e1f25799b6a3067141259c2c7b from #24208 which are specific for this issue, but it seems dependent on other changes to make a fix.

**Comment by jschwinger233** [comment]
For cilium_host binding with eth0 ipv6, do we want to make traffic host1 -> health@host2 ipsec-encrypted or not? This is not a problem for cilium_host binding with router ipv6, because curling from host uses router ipv6 as source ipv6, and we handle that as if it's from pod to pod.

There are an option `nodeEncryption` which controls ipsec encryption for node to node traffic, but I don't think host1 ->health@host2 has something to do with it.

I presume we want to let ipsec work, and the current situation is:
1. there is no xfrm policy for eth0 ipv6
2. health@host1 will response to ipv6_of_eth0@host1, but ipcache doesn't have correct entry for ipv6_of_eth0@host1 to set skb mark for further xfrm process;

I'm not 100% sure about how to handle this, do we want to add a masquerade to replace eth0 ipv6 with router ipv6? Or let xfrm policy as well as ipcache hold the right entry for eth0 ipv6? Adding a masquerade sounds like a easier way, shall I implement it using iptables?

**Comment by jschwinger233** [comment]
> Adding a masquerade sounds like a easier way, shall I implement it using iptables?

My mistake, we already have `encap_remap_v6_host_address` to do this.

**Comment by pchaigno** [comment]
> For cilium_host binding with eth0 ipv6, do we want to make traffic host1 -> health@host2 ipsec-encrypted or not? This is not a problem for cilium_host binding with router ipv6, because curling from host uses router ipv6 as source ipv6, and we handle that as if it's from pod to pod.

No, I don't think we want that. Any traffic that isn't pod-to-pod doesn't need to be encrypted. We sometimes encrypt hostns -> pod traffic today, but longer term, the plan is to stop doing that. It only causes issues because of the path asymmetry it implies (forward is encrypted, reply is not).

> There are an option `nodeEncryption` which controls ipsec encryption for node to node traffic, but I don't think host1 ->health@host2 has something to do with it.

That option was deprecated and we plan to remove it.

**Comment by jschwinger233** [comment]
>  Any traffic that isn't pod-to-pod doesn't need to be encrypted

Cool, thanks for clarification, this makes separation easier

**Comment by jschwinger233** [comment]
The season why this issue doesn't affect ipv4 is, traffic of local to health@remote uses router_ip@local as source ip, and lxc@remote which handles reply skb with dest ip router_ip@local will fetch an ipcache entry with an non-zero encryption key, then run into

https://github.com/cilium/cilium/blob/5c811513eae2bfe2bf745277f5c7b71b9290ca14/bpf/lib/encap.h#L208-L210

and mark the skb for ipsec. 

But that's not the case for ipv6, because the same response traffic on lxc@remote has eth0_ip@local for dest ip,  bpf program on lxc can't get an ipcache entry with non-zero encryption key and will instead reach

https://github.com/cilium/cilium/blob/5c811513eae2bfe2bf745277f5c7b71b9290ca14/bpf/lib/encap.h#L213-L227

For ipv6 scenario, the above code shows skb goes to kernel stack at 227L, and according to the context, I suppose we don't want this to happen, do we?

However, the comments there mention that we intentionally let skb go to stack for reversing NAT:

```
	/* For IPSec and the host firewall, traffic from a pod to a remote node
	 * is sent through the tunnel. In the case of node --> VIP@remote pod,
	 * packets may be DNATed when they enter the remote node. If kube-proxy
	 * is used, the response needs to go through the stack on the way to
	 * the tunnel, to apply the correct reverse DNAT.
	 * See #14674 for details.
```

So how can I distinguish different situations and make decision where a skb should go?

**Comment by jschwinger233** [comment]
BTW, the fix in #24208 makes ipv6 datapath the same as ipv4, which lets host-to-remote-pod traffic be ipsec-encrypted. But that solution doesn't work for cilium_host binding with eth0 ipv6, so we have to figure out a separate approach.

**Comment by brb** [comment]
@jschwinger233 You can drop this https://github.com/cilium/cilium-cli/blob/main/connectivity/tests/service.go#L220 :-)
